flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -336,1697 +386,1982 @@ public int getKeyGroupPrefixBytes() {
+						int arrayCmpRes = compareKeyGroupsForByteArrays(
+							o1.currentKey, o2.currentKey, currentBytes);
+						return arrayCmpRes == 0 ? o1.getKvStateId() - o2.getKvStateId() : arrayCmpRes;
+					}
+				});
+			}
+		}
+
+		RocksDBMergeIterator(List<Tuple2<RocksIterator, Integer>> kvStateIterators, final int keyGroupPrefixByteCount) {
+			Preconditions.checkNotNull(kvStateIterators);
+			this.keyGroupPrefixByteCount = keyGroupPrefixByteCount;
+
+			Comparator<MergeIterator> iteratorComparator = COMPARATORS.get(keyGroupPrefixByteCount);
+
+			if (kvStateIterators.size() > 0) {
+				PriorityQueue<MergeIterator> iteratorPriorityQueue =
+					new PriorityQueue<>(kvStateIterators.size(), iteratorComparator);
+
+				for (Tuple2<RocksIterator, Integer> rocksIteratorWithKVStateId : kvStateIterators) {
+					final RocksIterator rocksIterator = rocksIteratorWithKVStateId.f0;
+					rocksIterator.seekToFirst();
+					if (rocksIterator.isValid()) {
+						iteratorPriorityQueue.offer(new MergeIterator(rocksIterator, rocksIteratorWithKVStateId.f1));
+					} else {
+						IOUtils.closeQuietly(rocksIterator);
+					}
+				}
+
+				kvStateIterators.clear();
+
+				this.heap = iteratorPriorityQueue;
+				this.valid = !heap.isEmpty();
+				this.currentSubIterator = heap.poll();
+			} else {
+				// creating a PriorityQueue of size 0 results in an exception.
+				this.heap = null;
+				this.valid = false;
+			}
+
+			this.newKeyGroup = true;
+			this.newKVState = true;
 		}
 
 		/**
-		 * Restores all key-groups data that is referenced by the passed state handles.
-		 *
-		 * @param keyedStateHandles List of all key groups state handles that shall be restored.
+		 * Advance the iterator. Should only be called if {@link #isValid()} returned true. Valid can only chance after
+		 * calls to {@link #next()}.
 		 */
-		public void doRestore(Collection<KeyedStateHandle> keyedStateHandles)
-			throws IOException, StateMigrationException, RocksDBException {
+		public void next() {
+			newKeyGroup = false;
+			newKVState = false;
 
-			rocksDBKeyedStateBackend.createDB();
+			final RocksIterator rocksIterator = currentSubIterator.getIterator();
+			rocksIterator.next();
 
-			for (KeyedStateHandle keyedStateHandle : keyedStateHandles) {
-				if (keyedStateHandle != null) {
+			byte[] oldKey = currentSubIterator.getCurrentKey();
+			if (rocksIterator.isValid()) {
+				currentSubIterator.currentKey = rocksIterator.key();
 
-					if (!(keyedStateHandle instanceof KeyGroupsStateHandle)) {
-						throw new IllegalStateException("Unexpected state handle type, " +
-							"expected: " + KeyGroupsStateHandle.class +
-							", but found: " + keyedStateHandle.getClass());
-					}
-					this.currentKeyGroupsStateHandle = (KeyGroupsStateHandle) keyedStateHandle;
-					restoreKeyGroupsInStateHandle();
+				if (isDifferentKeyGroup(oldKey, currentSubIterator.getCurrentKey())) {
+					heap.offer(currentSubIterator);
+					currentSubIterator 
```

Comment:
I know this part is an old code block and wasn't be change in this PR. But it is a bit odds to me, kvStateInformation is not a copied object and it can be changed when writeKVStateMetaData()is invoking ... If I am not wrong, this is a serious bug. Am I misunderstand something?

Reply:
Strange sorry, my comment does not match the highlighted lines when line number > 1400, the problem I described above is on line: 1878 
```java
private void writeKVStateMetaData() throws IOException {
  // ...
        for (Map.Entry<String, Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>>> column :
				stateBackend.kvStateInformation.entrySet()) {
        }
  //...
}
```

Reply:
Yes, I remember that I fixed exactly this also in the heap backend. Defensive copy of the map in the sync part is the fix. I will just make this as a hotfix.

Reply:
Aha, I have make a PR for this ... Could you please have a look? [5525](https://github.com/apache/flink/pull/5525) I fix this in the way as you have pointed out.

Reply:
Thanks, but like I said, I have this already done in a hotfix that I will commit.

Reply:
Aha, closing it.